File: 1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#

# @lc code=start

# Solution #1: Brute Force
# Time Complexity: O(n^2)
# Space Complexity: O(1)
# class Solution:
#     def twoSum(self, nums: List[int], target: int) -> List[int]:
#         for i in range(len(nums)):
#             for j in range(i+1, len(nums)):
#                 if nums[i] + nums[j] == target:
#                     return [i, j]

# Solution #2: Use a hashmap
# Time Complexity: O(n)
# Space Complexity: O(n)
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashmap = {}

        for i, value in enumerate(nums):
            complement = target - value

            if complement in hashmap:
                return [hashmap[complement], i]

            hashmap[value] = i

# @lc code=end

File: 1.two-sum.ts
/*
 * @lc app=leetcode id=1 lang=typescript
 *
 * [1] Two Sum
 */

// @lc code=start

// Solution: Hashmap
// Time complexity: O(n)
// Space complexity: O(n)
function twoSum(nums: number[], target: number): number[] {
  const hashmap: Record<number, number> = {};

  for (let i = 0; i < nums.length; i++) {
    const value = nums[i];
    const complement = target - value;

    if (hashmap[complement] !== undefined) {
      return [hashmap[complement], i];
    }

    hashmap[value] = i;
  }

  return [];
}

// Test cases:
const test1 = twoSum([2, 7, 11, 15], 17);
console.log(test1);
// @lc code=end
